src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,612 @@
+module Crystal
+  class SourceTyper
+    # Represents a fully typed definition signature
+    record Signature,
+      name : String,
+      return_type : Crystal::ASTNode,
+      location : Crystal::Location,
+      args = {} of String => Crystal::ASTNode
+
+    getter program, files
+
+    def initialize(@entrypoint : String,
+                   @def_locators : Array(String),
+                   @excludes : Array(String),
+                   @type_blocks : Bool,
+                   @type_splats : Bool,
+                   @type_double_splats : Bool,
+                   @prelude : String = "prelude",
+                   stats : Bool = false,
+                   progress : Bool = false,
+                   error_trace : Bool = false)
+      @entrypoint = File.expand_path(@entrypoint) unless @entrypoint.starts_with?("/")
+      @program = Crystal::Program.new
+      @files = Set(String).new
+      @warnings = [] of String
+
+      @program.progress_tracker.stats = stats
+      @program.progress_tracker.progress = progress
+      @program.show_error_trace = error_trace
+    end
+
+    # Run the entire typing flow, from semantic to file reformatting
+    def run : Hash(String, String)
+      semantic(@entrypoint, File.read(@entrypoint))
+
+      rets = {} of String => String
+
+      @warnings.each do |warning|
+        puts "WARNING: #{warning}"
```

Comment:
Warnings are cases where type restrictions weren't added because it crossed a line with the language syntax somehow (i.e. empty splat calls)